ethereum/dataproviders/src/main/java/tech/pegasys/teku/dataproviders/lookup/BlockProvider.java

```diff
@@ -39,21 +38,18 @@ static BlockProvider fromDynamicMap(Supplier<Map<Bytes32, SignedBeaconBlock>> ma
   }
 
   static BlockProvider fromMapWithLock(
-      final Map<Bytes32, SignedBeaconBlock> blockMap,
-      final AsyncRunner asyncRunner,
-      final Lock readLock) {
-    return (roots) ->
-        asyncRunner.runAsync(
-            () -> {
-              readLock.lock();
-              try {
-                return roots.stream()
-                    .filter(blockMap::containsKey)
-                    .collect(Collectors.toMap(Function.identity(), blockMap::get));
-              } finally {
-                readLock.unlock();
-              }
-            });
+      final Map<Bytes32, SignedBeaconBlock> blockMap, final Lock readLock) {
+    return (roots) -> {
+      readLock.lock();
```

Comment:
i think potentially the answer should be more like fromMapWithLock becomes a member of the store, and the store doesn't go passing its locks about - it's kind of a bit of a scary concept...

If the store instead had 
```
private BlockProvider createBlockProvider(blockMap) {
    readLock.lock();
    try {
      return BlockProvider.fromMap(blockMap);
    } finally {
      readLock.unlock();
    }
}
```

Then we wouldn't be allowing other parties play with our read locks, which would be far cleaner, and also you can see where the read lock happens, without having to go digging down paths to other objects...

Also given this is called during constructor, it might even be better off locking before that process starts, and just using fromMap, as until this is done, it'd surely be bad to get stuck waiting for a lock... But I haven't thought that through fully... The upside would be queries to the store would be waiting on the lock...

Im still not convinced it's going to end well potentially - a safe future inside the constructor, if anything goes wrong the object creation (in this case the store) will fail... we might be better off rethinking the construction and having a create method where all of this is done up front and the constructor has a simpler job...  This might be more refactoring for now than we need, but worth considering @tbenr

Reply:
I didn't like sharing lock outside too, but the code you propose will do the lock when calling `createBlockProvider` instead of when using blockProvider. I don't see a way we could incorporate lock in BlockProvider without sharing it to BlockProvider or I don't understand something. The way we could go - drop `blocks` cache from BlockProvider at all and call `getBlockIfAvailable` along with BlockProvider whenever we need block from BlockProvider

Reply:
I agree that passing locks isn't a nice thing.. and it's confusing because sometime you don't know which lock you are dealing with. But a change here seems like out of scope and I'd like to be focused on restoring the status quo if this is the issue (and seems to be at least a part of the problem)